### PR TITLE
Authentication: members-only

### PR DESCRIPTION
**Pull request by Ceci007**
## members-only
In this project we:
- Add authentication with the devise gem.
- Use Bulma as our front-end framework.
- Add members-only and admin roles.

### CODE REVIEW

File: app/controllers/tweets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  # GET /tweets/new
+  def new
+    @tweet = Tweet.new
+    @tweet = current_user.tweets.build
+  end
+
+  # GET /tweets/1/edit
+  def edit; end
+
```
Comment (addod19) [inline app/controllers/tweets_controller.rb:28]: Let us remove edit and show method in our tweets controller and setup the routes correctly, like `resources :tweets, except[:edit, :show]`

File: app/controllers/tweets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
+class TweetsController < ApplicationController
+  before_action :set_tweet, only: %i[show edit update destroy]
+  before_action :authenticate_user!, except: %i[index show]
```
Comment (addod19) [inline app/controllers/tweets_controller.rb:3]: Our before action show restrict the `new` and `create` method

File: app/views/devise/mailer/email_changed.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
+  <p>We're contacting you to notify you that your email is being changed to <%= @resource.unconfirmed_email %>.</p>
+<% else %>
+  <p>We're contacting you to notify you that your email has been changed to <%= @resource.email %>.</p>
+<% end %>
```
Comment (addod19) [inline app/views/devise/mailer/email_changed.html.erb:7]: There should be no logic in views, instead, we could use the view helpers

File: app/views/devise/registrations/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+          <div class="field">
+            <div class="control">
+              <%= f.input :email, required: true, input_html: { class: "input"}, wrapper: false, label_html: { class: "label" } %>
+            </div>
+          </div>
+
+          <div class="field">
+          <% if devise_mapping.confirmable? && resource.pending_reconfirmation? %>
+            <p>Currently waiting confirmation for: <%= resource.unconfirmed_email %></p>
+          <% end %>
```
Comment (addod19) [inline app/views/devise/registrations/edit.html.erb:31]: There should be no logic in views, instead, we could use the view helpers

File: app/views/tweets/_feed.html.erb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        </div>
+      <% end %>
+      <div class="media-content">
+        <div class="content">
+        <% if user_signed_in? && (current_user.id == tweet.user_id || current_user.members_only? || current_user.admin?) %>
+          <strong><%= tweet.user.name %></strong><br/>
+          <small>@<%= tweet.user.username %></small><br/>
+        <% else %>
+          <strong>anonymous user</strong><br/>
+        <% end %>
```
Comment (addod19) [inline app/views/tweets/_feed.html.erb:37]: There should be no logic in views, instead, we could use the view helpers

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi @Ceci007 and @ABDELLANI-Youcef, great job on this project :100: The UI looks awesome, README is detailed with description :+1: However, there are some comments we need to address before we can progress. Kindly check the comments below:
- We are missing this functionality, `as a guest you see only post text`
pt
- `Routes` let us use only the resources used in our routes. For example, we did not use `show` and `edit` but we have declared them all in our routes.
- `Before action` should restrict our `new` and `create` method as in the screenshot below
bf
- Currently, we have logic in our views, let us put the in the view helpers

Kindly submit another code review once these are fixed :+1: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Request changes :recycle:
### Well done team :muscle: This is a good project and you have done a good job.
#### However, there are still logics in the view. Check my comments below. Fix these and submit them for another review.

#### _Review by Temitayo Adeleke_

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
#### There was a delay when I was reviewing the project so check the above comments for the required changes.

**Review by tirthajyoti-ghosh — APPROVED** [review]
# Approved ✅

## You have done an incredible job making this project 👏🧨💪! Just perfect 💯! 

### You have met all the requirements of the project. Go ahead and merge it! Onto the next project 🚀 !

Happy coding 👨‍💻 

<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Pull request merged**